compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -409,6 +409,103 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Produces a suggestion if it's highly likely that a user meant array
+    /// but accidentaly used braces, causing the code to be interpreted as
+    /// a block expression.
+    pub(super) fn maybe_suggest_brackets_instead_of_braces(&mut self, lo: Span) {
+        if lo.from_expansion() {
+            return;
+        }
+
+        let mut delims = Vec::new();
+        let mut position = 1;
+        let mut hi = DUMMY_SP;
+
+        // NOTE(hkmatsumoto): The initial value is asymmetric because of the
+        // behavior of `look_ahead`, which skips close delimiter while loads
+        // corresponding open delimiter. For example, we now have tokens like:
+        //
+        // `{(1, 2, 3), 2, 3}`
+        //   ^[1]     ^[2]
+        // (the outer-most braces represent block expression)
+        //
+        // `look_ahead(1, |t: &Token| { t })` outputs `(`[1], and
+        // `look_ahead(2, |t: &Token| { t })` outputs `,`[2].
+        //
+        // As a workaround, we do nothing for `token::OpenDelim(token::Brace)`
+        // in the match expression below, just waiting for
+        // `token::CloseDelim(token::Brace)` to exit the loop.
+        let mut depth = [
+            1, // {}
+            0, // <>, this appears in fn calls. e.g., {foo::<A, B>(), 2, 3}
+        ];
+
+        // We keep calling `look_ahead` until we reach the end of the block expression.
+        loop {
```

Comment:
I have a [somewhat different approach](https://github.com/rust-lang/rust/pull/88729) that might be able to make some of this code easier to reason about.

Instead of using `look_ahead` for cases where we kind of need to parse things with non-deterministic lookahead, I do `let snapthot = self.clone()` on the `Parser` and *try to successfully parse* what the "fake" syntax would be. If that succeeds, we know that we can provide an accurate suggestion, if that doesn't, we cancel the error and rollback the parser to the previous state to emit the "normal" error we currently see and let the existing error recovery machinery take its course. With this approach, it should be possible to emit a suggestion on a case like `{0, if 1 < 2 {1} else {1}, 2, 3}`. Because cloning is somewhat expensive (but not terrible, from what I've seen), we want to avoid it on the "happy path" of well-formed code, but you can use limited `look_ahead` to see if *it could be* this case, like doing `self.look_ahead(2, |t| t.kind == &Comma)` or something.

We could even try to keep information on the `Parser` for "we are in a `const` assignment, of array type" that we check here in `maybe_suggest_brackets_instead_of_braces`, and if it is not the case, we exit early, like we do for things coming from a macro.

What do you think? (Sorry for the delay in reviewing)

Reply:
It's very smart of you to come up with the approach :)
What's nice is that by cloning the parser the codebase will be significantly simplified while the suggestion keeps working almost as-is except some cases like `{if 1 < 2 {1} else {1}, 2, 3}`, but that's compromisable as I assume it's unlikely to initialize an array with complex expressions.

Reply:
Thank you, you're too kind.

Yes, I agree that this gets 80% there for 20% of the effort. If we were relying on this for the actual language and not just for error recovery, the story would be different.